dvc/remote/gs.py

```diff
@@ -138,16 +138,35 @@ def remove(self, path_info):
 
         blob.delete()
 
-    def _list_paths(self, bucket, prefix):
-        for blob in self.gs.bucket(bucket).list_blobs(prefix=prefix):
+    def _list_paths(self, path_info, max_items=None):
+        for blob in self.gs.bucket(path_info.bucket).list_blobs(
+            prefix=path_info.path, max_results=max_items
+        ):
             yield blob.name
 
     def list_cache_paths(self):
-        return self._list_paths(self.path_info.bucket, self.path_info.path)
+        return self._list_paths(self.path_info)
+
+    def walk_files(self, path_info):
+        for fname in self._list_paths(path_info):
+            # skip nested empty directories
+            if fname.endswith("/"):
+                continue
+            yield path_info.replace(fname)
+
+    def makedirs(self, path_info):
+        self.gs.bucket(path_info.bucket).blob(
+            (path_info / "").path
+        ).upload_from_string("")
+
+    def isdir(self, path_info):
+        dir_path = path_info / ""
+        return bool(list(self._list_paths(dir_path, max_items=1)))
 
     def exists(self, path_info):
-        paths = set(self._list_paths(path_info.bucket, path_info.path))
-        return any(path_info.path == path for path in paths)
+        dir_path = path_info / ""
+        fname = next(self._list_paths(path_info, max_items=1), "")
+        return path_info.path == fname or fname.startswith(dir_path.path)
```

Comment:
Just found out that this logic is wrong.

Say, we have a file `data/subdir-file.txt` and `data/subdir/1`. When we go add `data/subdir`, it will find substr for `subdir-file.txt`, but won't really find `subdir` as there's no check for `isdir`. And, therefore, it would fail with a message that the `'data/subdir' does not exist`.

Reply:
Same logic is in `S3`

Reply:
This line `dir_path = path_info / ""` adds a trailing `/` to `dir_path`, so `subdir-file.txt` won't start with `dir_path.path`.

Reply:
Yes, sorry, I did not explain it correctly. The `next()` could provide you `subdir-file.txt` which won't match, therefore, it returning False, resulting in `'data/subdir' does not exist` message.

There's no short-circuiting for checking if a given file/folder exists. You have to first check if the folder exists, and then, loop through it until you find exact match.